src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -98,7 +98,8 @@ internal class LocalGrainDirectory :
             ISiloStatusOracle siloStatusOracle,
             IMultiClusterOracle multiClusterOracle,
             IInternalGrainFactory grainFactory,
-            GrainTypeManager grainTypeManager)
+            Func<GrainDirectoryPartition> grainDirectoryPartitionFactory,
```

Comment:
I'm a little on the fence with using Func<T> as factory methods all throughout, as they don't really disambiguate whether it gets the T, or creates it. In this particular case, you explicitly want something that creates it.
Subjectively, to me it feels that Func<T> just retrieves the T (you might get a different one if config changes for example, but doesn't create a new transient object every single time it's called).
Still, in a practical sense, it would be equivalent, and avoids having to define `Factory<T>`, `Factory<T1,T2>`, etc.
I believe @jason-bragg had a similar concern and started a thread about it not long ago.
Thoughts?